diff --git a/res/values/config.xml b/res/values/config.xml
index f12ac5b..0aebe09 100644
--- a/res/values/config.xml
+++ b/res/values/config.xml
@@ -8,6 +8,6 @@
 -->
 <resources xmlns:xliff="urn:oasis:names:tc:xliff:document:1.2">
     <!-- CMUpdate Config Strings -->
-    <string name="conf_update_server_url_def" translatable="false">https://download.cyanogenmod.org/api</string>
+    <string name="conf_update_server_url_def" translatable="false">https://raw.githubusercontent.com/Verevka/API_CMUpdater/master/API</string>
     <string name="conf_changelog_url" translatable="false">http://localhost/changelog.xml</string>
 </resources>
diff --git a/src/com/cyanogenmod/updater/service/DownloadService.java b/src/com/cyanogenmod/updater/service/DownloadService.java
index f4b2b80..cb38faa 100644
--- a/src/com/cyanogenmod/updater/service/DownloadService.java
+++ b/src/com/cyanogenmod/updater/service/DownloadService.java
@@ -95,7 +95,7 @@ public class DownloadService extends IntentService
     }
 
     private UpdatesJsonObjectRequest buildRequest(String sourceIncremental) {
-        URI requestUri = URI.create(getServerUri() + "/v1/build/get_delta");
+        URI requestUri = URI.create(getServerUri());
         UpdatesJsonObjectRequest request;
 
         // Set request body
